v1/src/main/java/com/google/cloud/teleport/templates/common/CsvConverters.java

```diff
@@ -0,0 +1,306 @@
+                + "https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.Predefined.html")
+    @Default.String("Default")
+    ValueProvider<String> getCsvFormat();
+
+    void setCsvFormat(ValueProvider<String> csvFormat);
+
+    @TemplateParameter.Text(
+        order = 5,
+        optional = true,
+        regexes = {"^(US-ASCII|ISO-8859-1|UTF-8|UTF-16)$"},
+        description = "CSV file encoding",
+        helpText =
+            "CSV file character encoding format. Allowed Values are US-ASCII"
+                + ", ISO-8859-1, UTF-8, UTF-16")
+    @Default.String("UTF-8")
+    ValueProvider<String> getCsvFileEncoding();
+
+    void setCsvFileEncoding(ValueProvider<String> csvFileEncoding);
+  }
+
+  /**
+   * Default value factory to get delimiter from Csv format so that if the user does not pass one
+   * in, it matches the supplied {@link CsvPipelineOptions#getCsvFormat()}.
+   */
+  public static class DelimiterFactory implements DefaultValueFactory<ValueProvider<String>> {
+
+    @Override
+    public ValueProvider<String> create(PipelineOptions options) {
+      return DualInputNestedValueProvider.of(options.as(CsvPipelineOptions.class).getCsvFormat(),
+          ValueProvider.StaticValueProvider.of(null),
+          new SerializableFunction<TranslatorInput<String, String>, String>() {
+            @Override
+            public String apply(TranslatorInput<String, String> input) {
+              CSVFormat format = CSVFormat.Predefined.valueOf(input.getX()).getFormat();
+              return String.valueOf(format.getDelimiter());
+            }
+      });
+    }
+  }
+
+  /**
+   * The {@link ReadCsv} class is a {@link PTransform} that reads from one for more Csv files. The
+   * transform returns a {@link PCollectionTuple} consisting of the following {@link PCollection}:
+   *
+   * <ul>
+   *   <li>{@link ReadCsv#headerTag()} - Contains headers found in files if read with headers,
+   *       contains empty {@link PCollection} if no headers.
+   *   <li>{@link ReadCsv#lineTag()} - Contains Csv lines as a {@link PCollection} of strings.
+   * </ul>
+   */
+  @AutoValue
+  public abstract static class ReadCsv extends PTransform<PBegin, PCollectionTuple> {
+
+    public static Builder newBuilder() {
+      return new AutoValue_CsvConverters_ReadCsv.Builder();
+    }
+
+    public abstract ValueProvider<String> csvFormat();
+
+    @Nullable
+    public abstract ValueProvider<String> delimiter();
+
+    public abstract ValueProvider<Boolean> hasHeaders();
+
+    public abstract ValueProvider<String> inputFileSpec();
+
+    public abstract TupleTag<String> headerTag();
+
+    public abstract TupleTag<String> lineTag();
+
+    public abstract ValueProvider<String> fileEncoding();
+
+    @Override
+    public PCollectionTuple expand(PBegin input) {
+
+      if (hasHeaders().get()) {
```

Comment:
how to perform check since it's a value provider now?

Exception in thread "main" java.lang.IllegalStateException: Value only available at runtime, but accessed from a non-runtime context: RuntimeValueProvider{propertyName=containsHeaders, default=false}
	at org.apache.beam.sdk.options.ValueProvider$RuntimeValueProvider.get(ValueProvider.java:254)
	at com.google.cloud.teleport.templates.common.CsvConverters$ReadCsv.expand(CsvConverters.java:205)
	at com.google.cloud.teleport.templates.common.CsvConverters$ReadCsv.expand(CsvConverters.java:180)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:545)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:496)
	at org.apache.beam.sdk.values.PBegin.apply(PBegin.java:56)
	at org.apache.beam.sdk.Pipeline.apply(Pipeline.java:190)
	at com.google.cloud.teleport.templates.CSVToBigQuery.main(CSVToBigQuery.java:203)

Reply:
This can only be done inside a DoFn, not during pipeline/graph expansion

Reply:
I see, I tried move the logic to DoFn. However, the issue is that the output tags depends on the condition. When we set the output tags, it's already outside of DoFn. All the examples I find so far is always hard coded the output tags. Do you have any suggestions?

if (condition is true) {
            return input
            .apply("MatchFilePattern", FileIO.match().filepattern(inputFileSpec()))
            .apply("ReadMatches", FileIO.readMatches())
            .apply(
                "ReadCsvWithHeaders",
                ParDo.of(
                        new GetCsvHeadersFn(
                            headerTag(), lineTag(), csvFormat(), delimiter(), fileEncoding()))
                    .withOutputTags(headerTag(), TupleTagList.of(lineTag())));
}
else {
return PCollectionTuple.of(
          lineTag(), input.apply("ReadCsvWithoutHeaders", TextIO.read().from(inputFileSpec())));
}